src/util.rs

```diff
@@ -8,3 +8,25 @@ pub fn leak<T>(b: Arc<T>) -> &'static T {
         &*p
     }
 }
+
+/// A dummy value that we can give to panic to indicate that we tried
+/// to acquire a poisoned lock, e.g. during flat-combining. The
+/// intention is for this to be treated specially by the panic
+/// propagation code -- in cases where the code must pick a "best"
+/// panic to propagate, it can ignore values of this type.
+pub struct PoisonPanic;
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ThreadId {
+    addr: usize
+}
+
+impl ThreadId {
+    pub fn current() -> Self {
+        // The actual `ID` value is irrelevant.  We're just using its TLS
+        // address as a unique thread key, faster than a real thread-id call.
+        thread_local!{ static ID: bool = false }
+        ID.with(|id| ThreadId { addr: id as *const bool as usize })
```

Comment:
Yeah, not really relevant -- I originally planned to write a more general purpose thing that made use of `ThreadId`. I'll just drop this commit I think.